js/apps/system/_admin/aardvark/APP/frontend/js/views/applicationDetailView.js

```diff
@@ -192,7 +192,7 @@
             } else if (result) {
               info = (
                 '<p>Script results:</p>' +
-                '<pre>' + JSON.stringify(result, null, 2) + '</pre>'
+                '<pre>' + JSON.stringify(arangoHelper.escapeHtml(result), null, 2) + '</pre>'
```

Comment:
Shouldn't this be `arangoHelper.escapeHtml(JSON.stringify(result), null, 2)`?

Reply:
i.e. it should also escape `{"xss": "<script>alert('hi')</script>"}` correctly but to me it looks like this would result in that being converted to `"[object Object]"` (assuming `arangoHelper.escapeHtml` expects a string).

Reply:
> Shouldn't this be `arangoHelper.escapeHtml(JSON.stringify(result), null, 2)`?

Why? Result needs to be escaped as it is the input with potential html code. We could remove the stringify if any input is expected. But it will not harm. But sure, if you have a better approach just inline edit and push.

Reply:
What is the purpose of `JSON.stringify` here? I assumed it's so we can represent whatever arbitrary data the script returns, which might be a JSON object (i.e. whatever you assigned to `module.exports` in the script).

In that case running `arangoHelper.escapeHtml` on it would coerce it to a string, try to escape any HTML in that string and then pass the resulting escaped string to `JSON.stringify` which would generate a JSON string representation of that string, which is then inserted into the HTML.

The previous behavior would have been to pass the arbitrary data to `JSON.stringify`, which would generate a JSON string representation, which is then inserted into the HTML.

As I understand it the new could would always output `"[object Object]"` if the script returns an object instead of a string. Do I misunderstand what the code is doing?

Reply:
Looks like `arangoHelper.escapeHtml` actually doesn't even coerce its argument to a string, it just assumes it's a string and will blow up if it's not something with a `replace` method as all it does is call `str.replace(/&/g, "&#38;").replace(/"/g, "&#34;").replace(/'/g, "&#39;").replace(/</g, "&#60;")` on its argument.

Reply:
I'll add a stringify method in case of a non-string to the escapeHtml method. Then the behaviour should be correct.

Reply:
Don't we have it now String-escaped twice? (assume the Replace does not replace anything here)

```
> let a = {b: "foo"}
undefined
> c = String(JSON.stringify(a, null, 2)).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;').replace(/'/g, '&#39;');
'{\n  &quot;b&quot;: &quot;foo&quot;\n}'
> d = JSON.stringify(c, null, 2)
'"{\\n  &quot;b&quot;: &quot;foo&quot;\\n}"'
```

now d adds an additional set of "quotes", and re-escapes the line breaks.

I think the output of escapeHTML cannot be a JSON anymore, but only a string, without JSON meaning (quotes are replaced).